README
ATTENTION!
==========
DO NOT USE TIMEKPR FROM THE TRUNK BRANCH YET!
It's under heavy construction


About
=====
Keep control of computer usage

This program will track and control the computer usage of your user accounts. 
You can limit their daily usage based on a timed access duration and 
configure periods of day when they can or cannot log in.

Install
=======

Requirements
------------
- You might need to patch timekpr/dirs.py to match your system.
- You might need to enable access and time plugins in respective pam.d/ files

Dependencies
------------

Currently the project requires (expressed as Ubuntu packages):
 - python >= 2.5
 - setup.py requires python-distutils and python-distutils-extra
 - In order to run: python-gtk2, python-glade2, python-pyparsing, procps, 
   libnotify-bin, login, libpam-modules, gksu or kdesu

Installation
-------------
Execute:
sudo python setup.py install

Need more options?
python setup.py --help
python setup.py --help-commands

History
=======

 - This is a complete re-write in python and vast improvement from the original 
timekpr bash script: http://tinyurl.com/waggledanceblog
 - This project was born after Charles Jackson's (crjackson) request for a 
program that time-limits user accounts: http://ubuntuforums.org/showthread.php?t=887769
 - Even Nedberg initially began to improve the original bash script from Chris 
Jackson, Savvas Radević packaged it and soon began co-development.
 - timekpr is now in experimental mode, highly unstable. Expect changes and a 
release that will be more distributable.

Thanks
======

 Special thanks go to Charles Jackson (crjackson), who has contributed a lot by 
suggesting and brainstorming useful features, as well as reporting bugs. We 
would also like to thank all the people posting at that topic at ubuntuforums 
for their helpful suggestions, tests and bug reports!
 Also thanks to Nicolas Laurance, who has pointed us to the proper python 
documentation and coding style.
 Finally, a huge thank you to the translators and bug triagers/reporters! 
Since there are so many of you, there's a top contributors list:
https://launchpad.net/timekpr/+topcontributors


debian/changelog
timekpr (0.3.5~ppa1~ubuntu10) UNRELEASED; urgency=medium

  * fixes bug #1467654 (from original timekpr repo)

 -- Eduards Bezverhijs <edzis@inbox.lv>  Fri, 10 Jul 2015 13:01:35 +0300

timekpr (0.3.5~ppa1~precise~ubuntu9) precise; urgency=low

  * Fixed issue (reported by maarts in forums) with long usernames
  ** using "who" instead of "w"

 -- Eduards Bezverhijs <edzis@inbox.lv>  Tue, 26 May 2015 09:21:17 +0200

timekpr (0.3.5~ppa1~trusty~ubuntu8) trusty; urgency=low

  * Trying to support 15.04
  * Functionality changes
  ** New method of determining logged in users (using w instead of loginctl if available, will fall back to loginctl or users in case w is not available)
  ** Notifications will be critical all the time (they will show up in fullscreen windows all the time)
  * Administration changes
  ** Now it will be possible to administer timekpr using non-administrative account (using pkexec now)
  ** It's possible to administer Timekpr from appindicator/systray menu

 -- Eduards Bezverhijs <edzis@inbox.lv>  Sun, 26 April 2015 15:21:17 +0200

timekpr (0.3.5~ppa1~14.04~ubuntu6) trusty; urgency=low

  * Translation related changes
  ** Cleaned up non-translatable strings (like authors / GPL)
  ** Made timekpr.pot file up-to-date so people can start translation w/o me sending them .po files
  * Cleaned up older files / experiments

 -- Eduards Bezverhijs <edzis@inbox.lv>  Sun, 6 July 2014 00:53:11 +0200

timekpr (0.3.5~ppa1~14.04~ubuntu5) trusty; urgency=low

  * Translation related changes
  ** Couple of translation fixes in About page
  ** Timekpr launcher now uses translated names
  ** German translations fixes (thanks to Kai Springstein)
  * KeyError fixes in timekpr-gui (thanks to Frank Gerbig)

 -- Eduards Bezverhijs <edzis@inbox.lv>  Sun, 6 July 2014 00:53:11 +0200

timekpr (0.3.5~ppa1~14.04~ubuntu4) trusty; urgency=low

  * Translation related changes
  ** Modfied/simplified (from code perspective) translations
  ** Introduced plural forms in translations
  ** Added German translation (thanks to Kai Springstein)
  * version fix in GUI

 -- Eduards Bezverhijs <edzis@inbox.lv>  Sun, 5 July 2014 20:11:39 +0200

timekpr (0.3.5~ppa1~14.04~ubuntu3) trusty; urgency=low

  * Translation related changes
  ** Added Russian (myself) translation
  * minor changes in GUI

 -- Eduards Bezverhijs <edzis@inbox.lv>  Sun, 21 June 2014 20:11:39 +0200

timekpr (0.3.5~ppa1~14.04~ubuntu2) trusty; urgency=low

  * Translation related changes
  ** Old translations are not working, old translations dropped, info - http://ubuntuforums.org/showthread.php?t=2219675&page=2
  ** Made all gui strings translatable (including menu and indicator/client messages)
  ** Added Latvian, French, Finnish translations - thanks to myself (Latvian), Jean-Phillipe Lord (French) and Petja Touru (Finnish)
  * Added MATE support - supports MDM and client for MATE DE (tested using Mint 17)

 -- Eduards Bezverhijs <edzis@inbox.lv>  Sun, 21 June 2014 20:11:39 +0200

timekpr (0.3.5~ppa1~14.04~ubuntu1) trusty; urgency=low

  Indicator/client changes
  * Added immediate notifications in case time is added to account to inform user about change asap
  * Unified indicator and tray client into one module (new functionality are the same for both now)
  * New icons which reflect time status
  ** unlimited time = unlocked green
  ** time left > 10 mins = locked green
  ** time left < 10 mins = locked yellow
  ** time left < 3 mins = locked red
  * All client notifications should be translatable

 -- Eduards Bezverhijs <edzis@inbox.lv>  Fri, 9 May 2014 20:11:39 +0200

timekpr (0.3.4~ppa1~14.04~ubuntu1) trusty; urgency=low

  * Trying to support LXDM
  * Integrate patch which initializes notification (thanks goes to SanskritFritz, ArchLinux AUR Timekpr maintainer)

 -- Eduards Bezverhijs <edzis@inbox.lv>  Mon, 7 May 2014 20:11:39 +0200

timekpr (0.3.3~ppa1~14.04~ubuntu2) trusty; urgency=low

  * Adapted to work with 14.04 (no backport planned at this time, although it might work even on 12.04)
  * Added loginctl support (will fall back to "users" in case loginctl is not present)
  * User list is filtered against popular DM's (gdm, kdm, lxdm, lightdm)
  * Added support for lightdm (tested only on 14.04)
  * NEW appindicator to support ubuntu's indicators
  * Appindicator works different from standard client, I hope - better

 -- Eduards Bezverhijs <edzis@inbox.lv>  Mon, 21 Mar 2014 20:11:39 +0200

timekpr (0.3.2~ppa1~ubuntu2) jaunty; urgency=low

  * Client includes a deafult notification for unsupported DEs (LP: #486881)
  * Release reflects revision 194 from timekpr/stable

 -- Even Nedberg <even@undertow>  Wed, 09 Dec 2009 21:15:15 +0100

timekpr (0.3.1~ppa1~ubuntu6) jaunty; urgency=low

  * Release reflects revision 193 from lp:timekpr/stable

 -- Even Nedberg <even@undertow>  Mon, 09 Nov 2009 20:29:44 +0100

timekpr (0.3.1~ppa1~ubuntu5~beta1) jaunty; urgency=low

  * Fixed an error in timekpr-client.py 

 -- Even Nedberg <even@undertow>  Tue, 03 Nov 2009 21:28:24 +0100

timekpr (0.3.1~ppa1~ubuntu2~beta1) januty; urgency=low

  * Added some files that got lost in last build

 -- Even Nedberg <even@lateralus>  Sun, 01 Nov 2009 19:14:28 +0100

timekpr (0.3.1~ppa1~ubuntu1~beta1) jaunty; urgency=low  

  * Fixed notifications in KDE4, LP: #302774
  * Cleaned up code for notification string (Stephen Scott)

 -- Even Nedberg <even@lateralus>  Sun, 01 Nov 2009 13:22:33 +0100

timekpr (0.3.0~ppa1~ubuntu1) jaunty; urgency=low

  * Fixed a bug that would kick a unrestricted user who was logged in all day
  * Release includes files from revision 221

 -- Even Nedberg <even@lateralus>  Sat, 18 Apr 2009 09:24:35 +0200

timekpr (0.3.0~ppa1~ubuntu1~beta4) jaunty; urgency=low

  * timekpr-client will no longer display two or more notifications at the same time
  * Added Finish translation

 -- Even Nedberg <even@undertow>  Sun, 12 Apr 2009 20:09:28 +0200

timekpr (0.3.0~ppa1~ubuntu1~beta3) jaunty; urgency=low

  * Added manpage (LP: #302770)


timekpr-pam.py
#!/usr/bin/env python
import re
from time import strftime

#TODO: Test and make it a module

## Check/enable/disable to /etc/pam.d/gdm and /etc/pam.d/login

## COMMON
def getconfsection(conffile):
	#Argument: string
	#Returns the section of the timekpr in a file
	s = open(conffile).read()
	m = re.compile('## TIMEKPR START\n(.*)\n## TIMEKPR END', re.S).findall(s)
	if len(m) > 1: exit('Error: More then one timekpr sections found(?): "' + conffile + '"')
	return m[0]

## Read/Write access.conf
def parseaccessconf(f = '/etc/security/access.conf'):
	#Parses the timekpr section in access.conf
	#Returns a list with the disabled usernames: ['niania','wawa']
	s = getconfsection(f)
	m = re.compile('^-:([^:]+):ALL$',re.M).findall(s)
	#If no matches (or bad format?), m = []
	return m

def isuserlocked(u):
	#Argument: username
	#Checks if user is in access.conf
	#Returns: True/False
	try: i = parseaccessconf().index(u)
	except ValueError: i = -1
	if i < 0: return False
	return True

def unlockuser(u, f = '/etc/security/access.conf'):
	#Argument: username
	#Removes access.conf line of user
	#Returns True (even if user is not locked) or False (if not possible)
	if isuserlocked(u) is False: return True
	fn = open(f,'w+')
	s = fn.read()
	m = re.sub('(## TIMEKPR START\n.*)-:'+u+':ALL\n','\\1', s)
	try: fn.write(m)
	except: return False
	fn.close()
	return True

def lockuser(u, f = '/etc/security/access.conf'):
	#Argument: username
	#Removes access.conf line of user
	#Returns True (even if user is already locked) or False
	if isuserlocked(u) is True: return True
	fn = open(f,'w+')
	s = fn.read()
	m = re.sub('(## TIMEKPR END.*)', '-:'+u+':ALL\n\\1', s)
	try: fn.write(m)
	except: return False
	fn.close()
	return True

## Read/write time.conf
def hourize(n):
	#make 7 into 0700, or 22 into 2200
	if int(n) < 10: return '0%s00' % str(n)
	return '%s00' % str(n)

def converttimeline(hfrom,hto):
	#all same:
	mfrom = re.compile('^(?:(\d+) ){6}\\1').findall(' '.join(hfrom))
	mto = re.compile('^(?:(\d+) ){6}\\1').findall(' '.join(hto))
	#return Al0700-2400
	if mfrom and mto: return 'Al' + hourize(mfrom[0]) + '-' + hourize(mto[0])
	
	#or separately all days
	su = 'Su' + hourize(hfrom[0]) + '-' + hourize(hto[0])
	mo = 'Mo' + hourize(hfrom[1]) + '-' + hourize(hto[1])
	tu = 'Tu' + hourize(hfrom[2]) + '-' + hourize(hto[2])
	we = 'We' + hourize(hfrom[3]) + '-' + hourize(hto[3])
	th = 'Th' + hourize(hfrom[4]) + '-' + hourize(hto[4])
	fr = 'Fr' + hourize(hfrom[5]) + '-' + hourize(hto[5])
	sa = 'Sa' + hourize(hfrom[6]) + '-' + hourize(hto[6])
	#Escaping \|, will be used in regular expressions
	return ' \| '.join([su,mo,tu,we,th,fr,sa])

def mktimeconfline(uname,hfrom,hto): return '\*;\*;'+uname+';'+converttimeline(hfrom,hto) #Escaping \*, will be used in regular expressions
""" Example:
hfrom = ['7', '7', '7', '7', '7', '7', '7']
hto = ['22', '22', '22', '22', '22', '22', '22']
mktimeconfline("username",hfrom,hto)
"""

def adduserlimits(username,bftom,bto,f = '/etc/security/time.conf'):
	#Adds a line with the username and their time limits in time.conf
	#Returns True or False (if it can't write to file)
	line = mktimeconfline(username,bfrom,bto)
	fn = open(f,'w+')
	s = fn.read()
	m = re.sub('(## TIMEKPR END.*)',line+'\n\\1', s)
	try: fn.write(m)
	except: return False
	fn.close()
	return True

def removeuserlimits(username,f = '/etc/security/time.conf'):
	#Removes a line with the username in time.conf
	#Returns True or False (if it can't write to file)
	fn = open(f,'w+')
	s = fn.read()
	m = re.sub('(## TIMEKPR START\n.*)\*;\*;'+username+'\n','\\1', s)
	try: fn.write(m)
	except: return False
	fn.close()
	return True

def isuserlimited(u,f = '/etc/security/time.conf'):
	#Argument: username
	#Checks if user is in time.conf
	#Returns: True/False
	s = getconfsection(f)
	m = re.compile('^\*;\*;([^;]+);',re.M).findall(s)
	try: i = m.index(u)
	except ValueError: i = -1
	if i < 0: return False
	return True

def isuserlimitednow(u,f = '/etc/security/time.conf'):
	#Argument: username
	#Checks if username should be limited as defined in time.conf
	#If this is True and the user is logged in, they should be killed
	#Returns: True or False (even if user is not in time.conf)
	if isuserlimited(u) is False: return False
	s = getconfsection(f)
	m = re.compile('^\*;\*;'+u+';(.*)$',re.M).findall(s)
	
	today = int(time.strftime("%w"))
	hournow = int(time.strftime("%H"))
	
	#If Al (All days):
	x = re.match('Al(\d\d)00-(\d\d)00',m[0])
	if x:
		low = int(x.group(1)) #lowest limit
		high= int(x.group(2)) #highest limit
		if low <= hournow < high: return False
	else:
		d = re.split(' \| ',m[0])[today]
		z = re.match('\w\w(\d\d)00-(\d\d)00',d)
		low = int(z.group(1))
		high = int(z.group(2))
		if low <= hournow < high: return False
	return True

def strint(x): return str(int(x)) #makes '08' into '8' and '10' as '10'

def converttconf(tfrom,tto,mode=0):
	""" In short, it removes the unnecessary 0 and multiplies tfrom,tto if necessary
	If mode=0 (default), it converts tfrom = ['08','08','13','14','15','01','09'], tto = ['22','14','19','20','21','23','25'] into ['8','8','13','14','15','1','9'] and ['22','14','19','20','21','23','25'] respectively
	If mode=1, it converts tfrom = '08', tto = '22' into ['8','8','8','8','8','8','8'] and ['22','22','22','22','22','22','22'] respectively
	"""
	if mode == 0:
		ffrom = map(strint,tfrom)
		fto = map(strint,tto)
	elif mode == 1:
		#Single values mode, need to multiply 7 times
		ffrom = [strint(tfrom)]*7
		fto = [strint(tto)]*7
	return ffrom,fto

def parsetimeconf(f = '/etc/security/time.conf'):
	#Returns a list with usernames and from/to configurations from the time.conf file
	c = getconfsection(f)
	utlist = re.compile('^\*;\*;([^;]+);(.*)$', re.M).findall(c)
	return utlist
	#example: [('niania', 'Al0000-2400'),('wawa', 'Su0700-2200 | Mo0700-2200 | Tu0700-2200 | We0700-2200 | Th0700-2200 | Fr0700-2200 | Sa0900-2200')]

def parseutlist(utlist):
	#Parses the list from parsetimeconf()
	retlist = []
	for u,t in utlist:
		#('niania', 'Al0000-2400')
		#u = 'niania' and t = 'Al0000-2400'
		
		#Check if Al is used
		checkAl = re.compile('^Al(\d{2})00-(\d{2})00$').findall(t)
		if checkAl:
			final = converttconf(checkAl[0][0],checkAl[0][1],1)
		else:
			#Break time in pieces
			pieces = re.split(' \| ',t)
			if len(pieces) != 7: exit('Error: Unsupported format detected (should have 7 time items): "' + t + '"')
			if not re.search('^Su\d{2}00-\d{2}00$',pieces[0]): exit('Error: Unsupported format detected (Sunday should be first): "' + t + '"')
			#0=Sunday, su[0] is from, su[1] is to
			su = re.compile('^Su(\d\d)00-(\d\d)00$').findall(pieces[0])[0]
			mo = re.compile('^Mo(\d\d)00-(\d\d)00$').findall(pieces[1])[0]
			tu = re.compile('^Tu(\d\d)00-(\d\d)00$').findall(pieces[2])[0]
			we = re.compile('^We(\d\d)00-(\d\d)00$').findall(pieces[3])[0]
			th = re.compile('^Th(\d\d)00-(\d\d)00$').findall(pieces[4])[0]
			fr = re.compile('^Fr(\d\d)00-(\d\d)00$').findall(pieces[5])[0]
			sa = re.compile('^Sa(\d\d)00-(\d\d)00$').findall(pieces[6])[0]
			final = converttconf([su[0],mo[0],tu[0],we[0],th[0],fr[0],sa[0]],[su[1],mo[1],tu[1],we[1],th[1],fr[1],sa[1]])
		retlist.append((u,final))
		#Append example:
		#user: [niania,(['0', '0', '0', '0', '0', '0', '0'], ['24', '24', '24', '24', '24', '24', '24'])]
		#user: [wawa,(['7', '7', '7', '7', '7', '7', '9'], ['22', '22', '22', '22', '22', '22', '22'])]
	return retlist
	"""Returns a list (retlist):
	[0] = first item:
		[0] = username niania
		[1] = fromto:
			[0] = from ['0', '0', '0', '0', '0', '0', '0']
			[1] = to ['24', '24', '24', '24', '24', '24', '24']
	Example: parseutlist(utlist)[0][1][0]
		['0', '0', '0', '0', '0', '0', '0']
	"""
